recipes/full_finetune_distributed.py

```diff
@@ -199,21 +204,101 @@ def __init__(self, cfg: DictConfig) -> None:
+        return self._checkpointer.load_checkpoint()
+
+    def resume_from_checkpoint(self, checkpoint_dict: Dict[str, any]) -> Dict[str, Any]:
+        """
+        Load the intermediate checkpoint to restore the training state.
+        If async checkpointing is enabled, user does not need to provide any recipe state.
+        DistributedCheckpoint loads the latest intermediate checkpoint state which includes the recipe state already.
+        Otherwise, update the recipe state as loaded in the checkpoint already via the load_checkpoint call.
+        """
+        if self._enable_async_checkpointing:
+            if self._is_rank_zero:
+                dcp_load_start = time.perf_counter()
+
+            model_state_dict = self._model.state_dict()
+
+            if not self._optimizer_in_bwd:
+                _init_optim_state(self._optimizer)
+                optim_state_dict = self._optimizer.state_dict()
+            else:
+                optim_state_dict = self._optim_ckpt_wrapper.state_dict()
+
+            state_dict: Dict[str:Any] = {}
+            state_dict.update(
+                {
+                    training.MODEL_KEY: model_state_dict,
+                    training.OPT_KEY: optim_state_dict,
+                    training.SEED_KEY: 0,
+                    training.EPOCHS_KEY: 0,
+                    training.TOTAL_EPOCHS_KEY: 0,
+                    training.MAX_STEPS_KEY: 0,
+                }
+            )
+
+            dcp_checkpointer = self.get_dcp_checkpointer()
+            state_dict = dcp_checkpointer.load_checkpoint(state_dict)
+
+            if self._is_rank_zero:
+                log.info(
+                    f"DistributedCheckpointer loaded the checkpoint in {time.perf_counter() - dcp_load_start:.2f} seconds."
+                )
+
+            if not self._optimizer_in_bwd:
+                set_state_dict(
+                    self._model,
+                    self._optimizer,
+                    model_state_dict=model_state_dict,
+                    optim_state_dict=optim_state_dict,
+                )
+            else:
+                set_model_state_dict(
+                    model=self._model,
+                    model_state_dict=model_state_dict,
+                )
+                self._optim_ckpt_wrapper.load_state_dict(optim_state_dict)
+
+            # Update the recipe state from the loaded checkpoint
+            self._update_recipe_state(state_dict)
+            log_rank_zero(
+                log,
+                msg=f"Loaded intermediate distributed checkpoint for the epoch: {state_dict[training.EPOCHS_KEY]}",
+            )
+        else:
             self._update_recipe_state(checkpoint_dict)
-        return checkpoint_dict
+            log_rank_zero(
+                log,
+                msg=f"Loaded intermediate checkpoint for the epoch: {checkpoint_dict[training.EPOCHS_KEY]}",
+            )
 
     def _update_recipe_state(self, ckpt_dict: Dict[str, Any]) -> None:
```

Comment:
I thought about it. Challenge is that it would be a slightly bigger refactor. The recipe keys are defined in the training module which causes a circular dependency. We would need to refactor that out of training first and then create an util for this method. I will leave it as it is for now since its an existing method in the recipe.